experimental/framework/fluid-static/src/serviceAudience.ts

```diff
@@ -12,67 +12,91 @@ import { IServiceAudience, IServiceAudienceEvents, IMember } from "./types";
+   * actually produce a change in what getMembers returns).  It also allows us to provide the
+   * client details in the event which makes it easier to find that client connection in a map
+   * keyed on the userId and not clientId.
+   * This map will always be up-to-date in a removeMember event because it is set once at
+   * construction and in every addMember event.
+   * It is mapped clientId to M to be better work with what the IAudience event provides
+   */
+  protected lastMembers: Map<string, M> = new Map();
+
   constructor(
       protected readonly container: Container,
   ) {
     super();
     this.audience = container.audience;
 
-    // Consolidating both the addition/removal of members
-    this.audience.on("addMember", () => {
-      this.emit("membersChanged", this.getMembers());
-    });
+    // getMembers will assign lastMembers so the removeMember event has what it needs
+    // in case it would fire before getMembers otherwise gets called the first time
+    this.getMembers();
 
-    this.audience.on("removeMember", () => {
-      this.emit("membersChanged", this.getMembers());
+    this.audience.on("addMember", (clientId: string, details: IClient) => {
+      if (this.shouldIncludeAsMember(details)) {
+        const member = this.getMember(clientId);
+        this.emit("addMember", clientId, member);
+        this.emit("membersChanged");
+      }
     });
 
-    this.container.on("connected", () => {
-      this.emit("membersChanged", this.getMembers());
+    this.audience.on("removeMember", (clientId: string) => {
+      if (this.lastMembers.has(clientId)) {
+        this.emit("removeMember", clientId, this.lastMembers.get(clientId));
+        this.emit("membersChanged");
+      }
     });
   }
 
+  protected abstract createServiceMember(audienceMember: IClient): M;
+
   /**
    * @inheritdoc
+   * ServiceAudience includes only interactive clients in its provided members.
    */
-  public getMembers(): Map<string, IMember> {
-    const users = new Map<string, IMember>();
+  public getMembers(): Map<string, M> {
+    const users = new Map<string, M>();
+    const clientMemberMap = new Map<string, M>();
     // Iterate through the members and get the user specifics.
     this.audience.getMembers().forEach((member: IClient, clientId: string) => {
-      // Get all the current human members
-      if (member.details.capabilities.interactive) {
+      if (this.shouldIncludeAsMember(member)) {
         const userId = member.user.id;
         // Ensure we're tracking the user
         let user = users.get(userId);
         if (user === undefined) {
-            user = { userId, connections: [] };
-            users.set(userId, user);
+          user = this.createServiceMember(member);
+          users.set(userId, user);
         }
 
         // Add this connection to their collection
         user.connections.push({ id: clientId, mode: member.mode });
+        clientMemberMap.set(clientId, user);
```

Comment:
If additional clients come in for the same user, would the `user.connections` value set for the first client in `clientMemberMap` include the other clients by the time `getMembers` finishes? 
i.e. if user 1 has client A, B, and C, would clientMemberMap["A"] return a member object with A,B, and C as connections?

Reply:
I'm pretty sure the synchronous behavior of events would mean new clients that come in while `getMembers` is running will not be reflected in this instance of `clientMemberMap`.  Then once this finishes running and the backing `IAudience` is able to process that it got another member, it will emit another `addMember` event and we will make a new `clientMemberMap`.

Reply:
Sounds good! I wasn't specifically referring to clients coming in while this function is running. Just a scenario when there are multiple clients per user. Since each client will get processed sequentially, client A will be set on the clientMemberMap while the user object doesn't have client B & C in connections, since those will only be added in when client B & C are being processed.
It should be fine however if IAudience is going to get updated and remove the in-memory map requirement altogether

Reply:
Ooh I see what you mean now.  Yeah user is a reference so additional clients will be reflected as expected in entries to `clientMemberMap` that were added before, and `Map` isn't doing any deep copying of values on insertion.